File: packages/node/tests/unit/claims-mapper.test.ts
import { describe, expect, it } from "vitest";
import { mapClaimsToUser } from "../../src/providers/microsoft/claims-mapper.js";
import { validateMicrosoftConfig } from "../../src/config/validate.js";
import { AuthenticationError } from "../../src/errors/index.js";
import { type RawClaims } from "../../src/core/claims.js";
import { CLIENT_ID, TENANT_ID } from "../helpers/crypto.js";

const config = (overrides = {}) =>
  validateMicrosoftConfig({
    clientId: CLIENT_ID,
    tenantId: "common",
    fetch: () => Promise.reject(new Error("unused")),
    ...overrides,
  });

describe("mapClaimsToUser", () => {
  it("maps default Microsoft claims to a normalized user", () => {
    const claims: RawClaims = {
      oid: "user-oid",
      sub: "subject",
      name: "Ada Lovelace",
      preferred_username: "ada@contoso.com",
      tid: TENANT_ID,
      roles: ["admin", "editor"],
      scp: "User.Read Mail.Send",
    };
    const user = mapClaimsToUser(claims, config(), "microsoft");
    expect(user).toMatchObject({
      id: "user-oid",
      name: "Ada Lovelace",
      email: "ada@contoso.com",
      tenantId: TENANT_ID,
      roles: ["admin", "editor"],
      scopes: ["User.Read", "Mail.Send"],
      provider: "microsoft",
    });
    expect(Object.isFrozen(user.claims)).toBe(true);
  });

  it("falls back through the default email source list", () => {
    const user = mapClaimsToUser({ oid: "x", upn: "u@contoso.com" }, config(), "microsoft");
    expect(user.email).toBe("u@contoso.com");
  });

  it("prefers oid then sub for id", () => {
    expect(mapClaimsToUser({ sub: "s" }, config(), "microsoft").id).toBe("s");
    expect(mapClaimsToUser({ oid: "o", sub: "s" }, config(), "microsoft").id).toBe("o");
  });

  it("defaults roles and scopes to empty arrays", () => {
    const user = mapClaimsToUser({ oid: "x" }, config(), "microsoft");
    expect(user.roles).toEqual([]);
    expect(user.scopes).toEqual([]);
    expect(user.email).toBeUndefined();
  });

  it("honors custom claim mapping overrides", () => {
    const user = mapClaimsToUser(
      { custom_id: "cid", mail: "c@x.com", display: "C", groups: ["g1", "g2"] },
      config({
        claims: { id: "custom_id", email: "mail", name: "display", roles: "groups" },
      }),
      "microsoft",
    );
    expect(user).toMatchObject({
      id: "cid",
      email: "c@x.com",
      name: "C",
      roles: ["g1", "g2"],
    });
  });

  it("accepts a space-delimited string for roles via custom mapping", () => {
    const user = mapClaimsToUser(
      { oid: "x", wids: "role-a role-b" },
      config({ claims: { roles: "wids" } }),
      "microsoft",
    );
    expect(user.roles).toEqual(["role-a", "role-b"]);
  });

  it("throws when no usable subject is present", () => {
    expect(() => mapClaimsToUser({ name: "x" }, config(), "microsoft")).toThrow(
      AuthenticationError,
    );
  });
});

File: packages/go/provider/microsoft/claims_mapper_test.go
package microsoft

import (
	"testing"

	"github.com/avi-pathak/easy-sso/packages/go/config"
	"github.com/avi-pathak/easy-sso/packages/go/core"
	"github.com/avi-pathak/easy-sso/packages/go/ssoerr"
)

func TestMapClaimsDefaults(t *testing.T) {
	cfg := norm(t, config.MicrosoftAuthConfig{ClientID: "c", TenantID: "common"})
	claims := core.Claims{
		"oid":   "object-id",
		"sub":   "subject",
		"email": "user@example.com",
		"name":  "Ada Lovelace",
		"tid":   "tenant-1",
		"roles": []any{"admin", "reader", ""},
		"scp":   "read write",
	}
	u, err := MapClaimsToUser(claims, cfg, "microsoft")
	if err != nil {
		t.Fatalf("map: %v", err)
	}
	if u.ID != "object-id" { // oid wins over sub
		t.Fatalf("id = %q", u.ID)
	}
	if u.Email != "user@example.com" || u.Name != "Ada Lovelace" || u.TenantID != "tenant-1" {
		t.Fatalf("unexpected user: %+v", u)
	}
	if len(u.Roles) != 2 || u.Roles[0] != "admin" || u.Roles[1] != "reader" {
		t.Fatalf("roles = %v (empty strings should be filtered)", u.Roles)
	}
	if len(u.Scopes) != 2 || u.Scopes[0] != "read" || u.Scopes[1] != "write" {
		t.Fatalf("scopes = %v (space-delimited scp)", u.Scopes)
	}
	if u.Provider != "microsoft" {
		t.Fatalf("provider = %q", u.Provider)
	}
}

func TestMapClaimsFallbackToSub(t *testing.T) {
	cfg := norm(t, config.MicrosoftAuthConfig{ClientID: "c", TenantID: "common"})
	u, err := MapClaimsToUser(core.Claims{"sub": "only-sub"}, cfg, "microsoft")
	if err != nil {
		t.Fatalf("map: %v", err)
	}
	if u.ID != "only-sub" {
		t.Fatalf("id = %q, want sub fallback", u.ID)
	}
	if u.Roles == nil || u.Scopes == nil {
		t.Fatal("roles/scopes must be non-nil")
	}
}

func TestMapClaimsMissingSubject(t *testing.T) {
	cfg := norm(t, config.MicrosoftAuthConfig{ClientID: "c", TenantID: "common"})
	_, err := MapClaimsToUser(core.Claims{"email": "x@y.z"}, cfg, "microsoft")
	if ae, ok := ssoerr.As(err); !ok || ae.Code != ssoerr.CodeAuthentication {
		t.Fatalf("want authentication_error, got %v", err)
	}
}

func TestMapClaimsCustomMapping(t *testing.T) {
	cfg := norm(t, config.MicrosoftAuthConfig{
		ClientID: "c", TenantID: "common",
		Claims: config.ClaimMappingConfig{
			ID:    []string{"custom_id"},
			Roles: []string{"groups"},
		},
	})
	claims := core.Claims{
		"custom_id": "cid",
		"oid":       "ignored",
		"groups":    []any{"g1", "g2"},
	}
	u, err := MapClaimsToUser(claims, cfg, "microsoft")
	if err != nil {
		t.Fatalf("map: %v", err)
	}
	if u.ID != "cid" {
		t.Fatalf("id = %q, want custom mapping", u.ID)
	}
	if len(u.Roles) != 2 {
		t.Fatalf("roles = %v, want from groups", u.Roles)
	}
}
